File: python/src/azoth/eos/reference/hayduk_minhas_diffusivity.py
"""``eos.hayduk_minhas_diffusivity`` - the liquid binary diffusivity from the
Hayduk-Minhas correlation.

Spec: ``specs/calcs/eos/hayduk_minhas_diffusivity.toml``, which records the two
solvent correlations and the clamps.
"""

from __future__ import annotations

import math

from azoth._registry_gen import spec as _spec_for
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import HaydukMinhasDiffusivityResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning

CALC_ID = "eos.hayduk_minhas_diffusivity"


def hayduk_minhas_diffusivity(form: str, VA: Q, T: Q, eta: Q) -> HaydukMinhasDiffusivityResult:
    """The binary diffusion coefficient at infinite dilution, from Hayduk-Minhas.

    ``form`` selects the solvent correlation; ``VA`` is the solute molar volume at
    the normal boiling point and ``eta`` the solvent viscosity. ``VA`` and ``eta``
    are clamped to NeqSim's `[20, 500]` cm**3/mol and `[0.1, 100]` cP before the
    correlation.

    Args:
        form: ``"paraffin"`` or ``"aqueous"``.
        VA: solute molar volume at its normal boiling point.
        T: absolute temperature.
        eta: solvent dynamic viscosity at ``T``.

    Returns:
        The binary diffusion coefficient, in ``m**2/s``.

    Raises:
        OutOfRangeError: if ``T`` is not positive.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> va = q(4.0203262233375156e-5, "m**3/mol")
        >>> eta = q(9.163064908813372e-4, "Pa*s")
        >>> r = hayduk_minhas_diffusivity("paraffin", va, q(298.15, "K"), eta)
        >>> round(r.d.magnitude, 16)
        4.3917780890442e-09
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {
        "VA": input_to_si(spec, "VA", VA),
        "T": input_to_si(spec, "T", T),
        "eta": input_to_si(spec, "eta", eta),
    }
    apply_checks(checks.on_input, values.get, warnings)

    va_cm3 = min(500.0, max(20.0, values["VA"] * 1.0e6))
    eta_cp = min(100.0, max(0.1, values["eta"] * 1000.0))

    if form == "paraffin":
        d_cm2s = (
            13.3e-8
            * math.pow(values["T"], 1.47)
            * math.pow(eta_cp, 10.2 / va_cm3 - 0.791)
            / math.pow(va_cm3, 0.71)
        )
    else:
        vi_term = max(0.01, math.pow(va_cm3, -0.19) - 0.292)
        d_cm2s = (
            1.25e-8 * vi_term * math.pow(values["T"], 1.52) * math.pow(eta_cp, 9.58 / va_cm3 - 1.12)
        )
    d = d_cm2s * 1.0e-4

    apply_checks(checks.derived, lambda name: d if name == "d" else None, warnings)

    return HaydukMinhasDiffusivityResult(d=from_si(d, "m**2/s"), warnings=tuple(warnings))

File: crates/azoth-eos/src/hayduk_minhas_diffusivity.rs
//! `eos.hayduk_minhas_diffusivity` - the liquid binary diffusivity from the
//! Hayduk-Minhas correlation.
//!
//! Spec: `specs/calcs/eos/hayduk_minhas_diffusivity.toml`, which records the two
//! solvent correlations and the clamps.

use azoth_core::units::{
    DynamicViscosity, MolarVolume, ThermodynamicTemperature, square_meters_per_second,
};
use azoth_core::{Result, apply_checks};

use crate::results::HaydukMinhasDiffusivityResult;
use crate::spec_gen;

/// Which Hayduk-Minhas solvent correlation to apply.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HaydukMinhasForm {
    /// The hydrocarbon-solvent correlation.
    Paraffin,
    /// The water-rich-solvent correlation.
    Aqueous,
}

impl HaydukMinhasForm {
    /// The spec's spelling.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Paraffin => "paraffin",
            Self::Aqueous => "aqueous",
        }
    }
}

impl std::str::FromStr for HaydukMinhasForm {
    type Err = String;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        match s {
            "paraffin" => Ok(Self::Paraffin),
            "aqueous" => Ok(Self::Aqueous),
            other => Err(format!(
                "unknown Hayduk-Minhas form `{other}`; expected `paraffin` or `aqueous`"
            )),
        }
    }
}

/// The binary diffusion coefficient at infinite dilution, from the Hayduk-Minhas
/// correlation.
///
/// `form` selects the solvent correlation; `VA` is the solute molar volume at its
/// normal boiling point and `eta` the solvent viscosity. `VA` is clamped to `[20,
/// 500]` cm**3/mol and `eta` to `[0.1, 100]` cP before the correlation is applied.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `T` is not positive.
///
/// # Example
/// ```
/// use azoth_core::units::{cubic_meters_per_mole, kelvins, pascal_seconds};
/// use azoth_eos::{HaydukMinhasForm, hayduk_minhas_diffusivity};
///
/// let r = hayduk_minhas_diffusivity(HaydukMinhasForm::Paraffin,
///     cubic_meters_per_mole(4.0203262233375156e-5), kelvins(298.15),
///     pascal_seconds(9.163064908813372e-4))?;
/// assert!((r.d.value - 4.391778089044218e-9).abs() < 1e-15);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
#[allow(non_snake_case)] // `VA` and `T` are the symbols in the published equation
pub fn hayduk_minhas_diffusivity(
    form: HaydukMinhasForm,
    VA: MolarVolume,
    T: ThermodynamicTemperature,
    eta: DynamicViscosity,
) -> Result<HaydukMinhasDiffusivityResult> {
    let spec = &spec_gen::HAYDUK_MINHAS_DIFFUSIVITY_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "VA" => Some(VA.value),
            "T" => Some(T.value),
            "eta" => Some(eta.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let va_cm3 = (VA.value * 1.0e6).clamp(20.0, 500.0);
    let eta_cp = (eta.value * 1000.0).clamp(0.1, 100.0);

    let d_cm2s = match form {
        HaydukMinhasForm::Paraffin => {
            13.3e-8 * T.value.powf(1.47) * eta_cp.powf(10.2 / va_cm3 - 0.791) / va_cm3.powf(0.71)
        }
        HaydukMinhasForm::Aqueous => {
            let vi_term = (va_cm3.powf(-0.19) - 0.292).max(0.01);
            1.25e-8 * vi_term * T.value.powf(1.52) * eta_cp.powf(9.58 / va_cm3 - 1.12)
        }
    };
    let d = d_cm2s * 1.0e-4;

    apply_checks(
        spec.derived_checks(),
        |name| (name == "d").then_some(d),
        &mut warnings,
    )?;

    Ok(HaydukMinhasDiffusivityResult {
        d: square_meters_per_second(d),
        warnings,
    })
}
